xmodule/capa/xqueue_submission.py

```diff
@@ -0,0 +1,112 @@
+    """Interface to the external grading system."""
+
+    def __init__(self, block):
+        self.block = block
+
+    def _parse_json(self, data, name):
+        """Helper function to parse JSON safely."""
+        try:
+            return json.loads(data) if isinstance(data, str) else data
+        except json.JSONDecodeError as e:
+            raise JSONParsingError(name, str(e)) from e
+
+    def get_submission_params(self, header, payload):
+        """
+        Extracts student submission data from the given header and payload.
+        """
+        header = self._parse_json(header, "header")
+        payload = self._parse_json(payload, "payload")
+
+        queue_name = header.get('queue_name', 'default')
+
+        if not self.block:
+            raise GetSubmissionParamsError()
+
+        course_id = str(self.block.scope_ids.usage_id.context_key)
+        item_type = self.block.scope_ids.block_type
+        block_id = self.block.scope_ids.def_id.block_id
+        score = self.block.max_score()
+
+        item_id = f"block-v1:{course_id.replace('course-v1:', '')}+type@{item_type}+block@{block_id}"
+
+        try:
+            grader_payload = self._parse_json(payload["grader_payload"], "grader_payload")
+            grader = grader_payload.get("grader", '')
+        except KeyError as e:
+            raise MissingKeyError("grader_payload") from e
+
+        student_info = self._parse_json(payload["student_info"], "student_info")
+        student_id = student_info.get("anonymous_student_id")
+
+        if not student_id:
+            raise ValidationError("The field 'anonymous_student_id' is missing from student_info.")
+
+        student_answer = payload.get("student_response")
+        if student_answer is None:
+            raise ValidationError("The field 'student_response' does not exist.")
+
+        student_dict = {
+            'item_id': item_id,
+            'item_type': item_type,
+            'course_id': course_id,
+            'student_id': student_id
+        }
+
+        return student_dict, student_answer, queue_name, grader, score
+
+    def send_to_submission(self, header, body, files_to_upload=None):
+        """
+        Submits the extracted student data to the edx-submissions system.
+        """
+        from submissions.api import create_submission
+
+        try:
+            student_item, answer, queue_name, grader, score = self.get_submission_params(header, body)
+            return create_submission(  # pylint: disable=unexpected-keyword-arg
+                student_item,
+                answer,
+                queue_name=queue_name,
+                grader=grader,
+                score=score,
+                file=files_to_upload
+            )
+        except JSONParsingError as e:
+            log.error("%s", e)
+            return {"error": str(e)}
+
+        except MissingKeyError as e:
+            log.error("%s", e)
+            return {"error": str(e)}
+
+        except ValidationError as e:
```

Comment:
These three exceptions do the same thing. change to catch all 3 exceptions in the same clause. See https://docs.python.org/3/tutorial/errors.html#handling-exceptions